File: 1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        HashSet<Node> set = new HashSet<>();
        set.add(p);
        set.add(q);
        Node parent = p.parent;
        while (parent != null) {
            if (!set.add(parent)) {
                return parent;
            }
            parent = parent.parent;
        }
        parent = q.parent;
        while (parent != null) {
            if (!set.add(parent)) {
                return parent;
            }
            parent = parent.parent;
        }
        return null;
    }
}

File: 1790-lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        m = p
        n = q
        while m != n:
            m = m.parent if m.parent else q
            n = n.parent if n.parent else p
        return m